// @azure/arm-compute/dist/commonjs/operations/galleryScriptVersions.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GalleryScriptVersionsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing GalleryScriptVersions operations. */
class GalleryScriptVersionsImpl {
    client;
    /**
     * Initialize a new instance of the class GalleryScriptVersions class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List gallery Script Versions in a gallery Script Definition.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param options The options parameters.
     */
    listByGalleryScript(resourceGroupName, galleryName, galleryScriptName, options) {
        const iter = this.listByGalleryScriptPagingAll(resourceGroupName, galleryName, galleryScriptName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByGalleryScriptPagingPage(resourceGroupName, galleryName, galleryScriptName, options, settings);
            },
        };
    }
    async *listByGalleryScriptPagingPage(resourceGroupName, galleryName, galleryScriptName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listByGalleryScript(resourceGroupName, galleryName, galleryScriptName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listByGalleryScriptNext(resourceGroupName, galleryName, galleryScriptName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listByGalleryScriptPagingAll(resourceGroupName, galleryName, galleryScriptName, options) {
        for await (const page of this.listByGalleryScriptPagingPage(resourceGroupName, galleryName, galleryScriptName, options)) {
            yield* page;
        }
    }
    /**
     * List gallery Script Versions in a gallery Script Definition.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param options The options parameters.
     */
    _listByGalleryScript(resourceGroupName, galleryName, galleryScriptName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, galleryScriptName, options }, listByGalleryScriptOperationSpec);
    }
    /**
     * Custom ArmResourceRead operation template with CloudError as Error
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param options The options parameters.
     */
    get(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            galleryName,
            galleryScriptName,
            galleryScriptVersionName,
            options,
        }, getOperationSpec);
    }
    /**
     * Create or update a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param galleryScriptVersion Parameters supplied to the create or update gallery Script Version
     *                             operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                galleryScriptName,
                galleryScriptVersionName,
                galleryScriptVersion,
                options,
            },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Create or update a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param galleryScriptVersion Parameters supplied to the create or update gallery Script Version
     *                             operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options);
        return poller.pollUntilDone();
    }
    /**
     * Update a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param galleryScriptVersion Parameters supplied to the update gallery Script Version operation.
     * @param options The options parameters.
     */
    async beginUpdate(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                galleryScriptName,
                galleryScriptVersionName,
                galleryScriptVersion,
                options,
            },
            spec: updateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Update a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param galleryScriptVersion Parameters supplied to the update gallery Script Version operation.
     * @param options The options parameters.
     */
    async beginUpdateAndWait(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options) {
        const poller = await this.beginUpdate(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, galleryScriptVersion, options);
        return poller.pollUntilDone();
    }
    /**
     * Delete a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                galleryScriptName,
                galleryScriptVersionName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Delete a gallery Script Version.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param galleryScriptVersionName The name of the gallery Script Version to be retrieved.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, options) {
        const poller = await this.beginDelete(resourceGroupName, galleryName, galleryScriptName, galleryScriptVersionName, options);
        return poller.pollUntilDone();
    }
    /**
     * ListByGalleryScriptNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param galleryName The name of the Shared Image Gallery.
     * @param galleryScriptName The name of the gallery Script Definition to be retrieved.
     * @param nextLink The nextLink from the previous successful call to the ListByGalleryScript method.
     * @param options The options parameters.
     */
    _listByGalleryScriptNext(resourceGroupName, galleryName, galleryScriptName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, galleryScriptName, nextLink, options }, listByGalleryScriptNextOperationSpec);
    }
}
exports.GalleryScriptVersionsImpl = GalleryScriptVersionsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listByGalleryScriptOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/scripts/{galleryScriptName}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryScriptVersionList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/scripts/{galleryScriptName}/versions/{galleryScriptVersionName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
        Parameters.galleryScriptVersionName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/scripts/{galleryScriptName}/versions/{galleryScriptVersionName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        201: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        202: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        204: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryScriptVersion,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
        Parameters.galleryScriptVersionName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const updateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/scripts/{galleryScriptName}/versions/{galleryScriptVersionName}",
    httpMethod: "PATCH",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        201: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        202: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        204: {
            bodyMapper: Mappers.GalleryScriptVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryScriptVersion1,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
        Parameters.galleryScriptVersionName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/scripts/{galleryScriptName}/versions/{galleryScriptVersionName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.GalleryScriptVersionsDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.GalleryScriptVersionsDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.GalleryScriptVersionsDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.GalleryScriptVersionsDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
        Parameters.galleryScriptVersionName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByGalleryScriptNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryScriptVersionList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.nextLink,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.galleryScriptName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
