// @azure/arm-network/dist/commonjs/operations/adminRules.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminRulesImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing AdminRules operations. */
class AdminRulesImpl {
    client;
    /**
     * Initialize a new instance of the class AdminRules class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List all network manager security configuration admin rules.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options)) {
            yield* page;
        }
    }
    /**
     * List all network manager security configuration admin rules.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            configurationName,
            ruleCollectionName,
            options,
        }, listOperationSpec);
    }
    /**
     * Gets a network manager security configuration admin rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param ruleName The name of the rule.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, ruleName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            configurationName,
            ruleCollectionName,
            ruleName,
            options,
        }, getOperationSpec);
    }
    /**
     * Creates or updates an admin rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param ruleName The name of the rule.
     * @param adminRule The admin rule to create or update
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, ruleName, adminRule, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            configurationName,
            ruleCollectionName,
            ruleName,
            adminRule,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Deletes an admin rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param ruleName The name of the rule.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, ruleName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                networkManagerName,
                configurationName,
                ruleCollectionName,
                ruleName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes an admin rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param ruleName The name of the rule.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, ruleName, options) {
        const poller = await this.beginDelete(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, ruleName, options);
        return poller.pollUntilDone();
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param configurationName The name of the network manager Security Configuration.
     * @param ruleCollectionName The name of the network manager security Configuration rule collection.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkManagerName, configurationName, ruleCollectionName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            configurationName,
            ruleCollectionName,
            nextLink,
            options,
        }, listNextOperationSpec);
    }
}
exports.AdminRulesImpl = AdminRulesImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/securityAdminConfigurations/{configurationName}/ruleCollections/{ruleCollectionName}/rules",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.AdminRuleListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion, Parameters.top, Parameters.skipToken1],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.configurationName,
        Parameters.ruleCollectionName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/securityAdminConfigurations/{configurationName}/ruleCollections/{ruleCollectionName}/rules/{ruleName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.BaseAdminRule,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.configurationName,
        Parameters.ruleCollectionName1,
        Parameters.ruleName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/securityAdminConfigurations/{configurationName}/ruleCollections/{ruleCollectionName}/rules/{ruleName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.BaseAdminRule,
        },
        201: {
            bodyMapper: Mappers.BaseAdminRule,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    requestBody: Parameters.adminRule,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.configurationName,
        Parameters.ruleCollectionName1,
        Parameters.ruleName1,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/securityAdminConfigurations/{configurationName}/ruleCollections/{ruleCollectionName}/rules/{ruleName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        201: {},
        202: {},
        204: {},
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion, Parameters.force],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.configurationName,
        Parameters.ruleCollectionName1,
        Parameters.ruleName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.AdminRuleListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkManagerName1,
        Parameters.configurationName,
        Parameters.ruleCollectionName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
