// @azure/arm-network/dist/commonjs/operations/applicationGatewayWafDynamicManifests.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationGatewayWafDynamicManifestsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing ApplicationGatewayWafDynamicManifests operations. */
class ApplicationGatewayWafDynamicManifestsImpl {
    client;
    /**
     * Initialize a new instance of the class ApplicationGatewayWafDynamicManifests class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets the regional application gateway waf manifest.
     * @param location The region where the nrp are located at.
     * @param options The options parameters.
     */
    list(location, options) {
        const iter = this.getPagingAll(location, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.getPagingPage(location, options, settings);
            },
        };
    }
    async *getPagingPage(location, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._get(location, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._getNext(location, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *getPagingAll(location, options) {
        for await (const page of this.getPagingPage(location, options)) {
            yield* page;
        }
    }
    /**
     * Gets the regional application gateway waf manifest.
     * @param location The region where the nrp are located at.
     * @param options The options parameters.
     */
    _get(location, options) {
        return this.client.sendOperationRequest({ location, options }, getOperationSpec);
    }
    /**
     * GetNext
     * @param location The region where the nrp are located at.
     * @param nextLink The nextLink from the previous successful call to the Get method.
     * @param options The options parameters.
     */
    _getNext(location, nextLink, options) {
        return this.client.sendOperationRequest({ location, nextLink, options }, getNextOperationSpec);
    }
}
exports.ApplicationGatewayWafDynamicManifestsImpl = ApplicationGatewayWafDynamicManifestsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Network/locations/{location}/applicationGatewayWafDynamicManifests",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationGatewayWafDynamicManifestResultList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.$host, Parameters.subscriptionId, Parameters.location],
    headerParameters: [Parameters.accept],
    serializer,
};
const getNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ApplicationGatewayWafDynamicManifestResultList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.location,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
