// @azure/arm-network/dist/commonjs/operations/networkSecurityPerimeterAccessRules.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkSecurityPerimeterAccessRulesImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing NetworkSecurityPerimeterAccessRules operations. */
class NetworkSecurityPerimeterAccessRulesImpl {
    client;
    /**
     * Initialize a new instance of the class NetworkSecurityPerimeterAccessRules class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists the NSP access rules in the specified NSP profile.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkSecurityPerimeterName, profileName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkSecurityPerimeterName, profileName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, profileName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkSecurityPerimeterName, profileName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkSecurityPerimeterName, profileName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkSecurityPerimeterName, profileName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkSecurityPerimeterName, profileName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, profileName, options)) {
            yield* page;
        }
    }
    /**
     * Gets the specified NSP access rule by name.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param accessRuleName The name of the NSP access rule.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkSecurityPerimeterName, profileName, accessRuleName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            profileName,
            accessRuleName,
            options,
        }, getOperationSpec);
    }
    /**
     * Creates or updates a network access rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param accessRuleName The name of the NSP access rule.
     * @param parameters Parameters that hold the NspAccessRule resource to be created/updated.
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, networkSecurityPerimeterName, profileName, accessRuleName, parameters, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            profileName,
            accessRuleName,
            parameters,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Deletes an NSP access rule.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param accessRuleName The name of the NSP access rule.
     * @param options The options parameters.
     */
    delete(resourceGroupName, networkSecurityPerimeterName, profileName, accessRuleName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            profileName,
            accessRuleName,
            options,
        }, deleteOperationSpec);
    }
    /**
     * Lists the NSP access rules in the specified NSP profile.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkSecurityPerimeterName, profileName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, profileName, options }, listOperationSpec);
    }
    /**
     * Reconcile NSP access rules
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param accessRuleName The name of the NSP access rule.
     * @param parameters Parameters for NSP access rule reconcile
     * @param options The options parameters.
     */
    reconcile(resourceGroupName, networkSecurityPerimeterName, profileName, accessRuleName, parameters, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            profileName,
            accessRuleName,
            parameters,
            options,
        }, reconcileOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param profileName The name of the NSP profile.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkSecurityPerimeterName, profileName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            profileName,
            nextLink,
            options,
        }, listNextOperationSpec);
    }
}
exports.NetworkSecurityPerimeterAccessRulesImpl = NetworkSecurityPerimeterAccessRulesImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/profiles/{profileName}/accessRules/{accessRuleName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspAccessRule,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
        Parameters.accessRuleName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/profiles/{profileName}/accessRules/{accessRuleName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.NspAccessRule,
        },
        201: {
            bodyMapper: Mappers.NspAccessRule,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters49,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
        Parameters.accessRuleName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/profiles/{profileName}/accessRules/{accessRuleName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
        Parameters.accessRuleName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/profiles/{profileName}/accessRules",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspAccessRuleListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion, Parameters.top, Parameters.skipToken1],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const reconcileOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/profiles/{profileName}/accessRules/{accessRuleName}/reconcile",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: {
                type: { name: "Dictionary", value: { type: { name: "any" } } },
            },
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters50,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
        Parameters.accessRuleName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspAccessRuleListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkSecurityPerimeterName,
        Parameters.profileName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
