// @azure/arm-network/dist/commonjs/operations/reachabilityAnalysisIntents.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReachabilityAnalysisIntentsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing ReachabilityAnalysisIntents operations. */
class ReachabilityAnalysisIntentsImpl {
    client;
    /**
     * Initialize a new instance of the class ReachabilityAnalysisIntents class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets list of Reachability Analysis Intents .
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkManagerName, workspaceName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkManagerName, workspaceName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkManagerName, workspaceName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkManagerName, workspaceName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkManagerName, workspaceName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkManagerName, workspaceName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkManagerName, workspaceName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkManagerName, workspaceName, options)) {
            yield* page;
        }
    }
    /**
     * Gets list of Reachability Analysis Intents .
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkManagerName, workspaceName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, workspaceName, options }, listOperationSpec);
    }
    /**
     * Get the Reachability Analysis Intent.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisIntentName Reachability Analysis Intent name.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisIntentName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            reachabilityAnalysisIntentName,
            options,
        }, getOperationSpec);
    }
    /**
     * Creates Reachability Analysis Intent.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisIntentName Reachability Analysis Intent name.
     * @param body Reachability Analysis Intent object to create/update.
     * @param options The options parameters.
     */
    create(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisIntentName, body, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            reachabilityAnalysisIntentName,
            body,
            options,
        }, createOperationSpec);
    }
    /**
     * Deletes Reachability Analysis Intent.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisIntentName Reachability Analysis Intent name.
     * @param options The options parameters.
     */
    delete(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisIntentName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            reachabilityAnalysisIntentName,
            options,
        }, deleteOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkManagerName, workspaceName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            nextLink,
            options,
        }, listNextOperationSpec);
    }
}
exports.ReachabilityAnalysisIntentsImpl = ReachabilityAnalysisIntentsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisIntents",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisIntentListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.skipToken,
        Parameters.skip,
        Parameters.top1,
        Parameters.sortKey,
        Parameters.sortValue,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisIntents/{reachabilityAnalysisIntentName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisIntent,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisIntentName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisIntents/{reachabilityAnalysisIntentName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisIntent,
        },
        201: {
            bodyMapper: Mappers.ReachabilityAnalysisIntent,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    requestBody: Parameters.body3,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisIntentName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisIntents/{reachabilityAnalysisIntentName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisIntentName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisIntentListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
