// @azure/arm-network/dist/commonjs/operations/serviceGateways.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceGatewaysImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing ServiceGateways operations. */
class ServiceGatewaysImpl {
    client;
    /**
     * Initialize a new instance of the class ServiceGateways class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets all the service gateways in a subscription.
     * @param options The options parameters.
     */
    listAll(options) {
        const iter = this.listAllPagingAll(options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listAllPagingPage(options, settings);
            },
        };
    }
    async *listAllPagingPage(options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listAll(options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listAllNext(continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listAllPagingAll(options) {
        for await (const page of this.listAllPagingPage(options)) {
            yield* page;
        }
    }
    /**
     * Gets all the service gateways in a resource group.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param options The options parameters.
     */
    list(resourceGroupName, options) {
        const iter = this.listPagingAll(resourceGroupName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, options)) {
            yield* page;
        }
    }
    /**
     * Get address locations in service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    listAddressLocations(resourceGroupName, serviceGatewayName, options) {
        const iter = this.getAddressLocationsPagingAll(resourceGroupName, serviceGatewayName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.getAddressLocationsPagingPage(resourceGroupName, serviceGatewayName, options, settings);
            },
        };
    }
    async *getAddressLocationsPagingPage(resourceGroupName, serviceGatewayName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._getAddressLocations(resourceGroupName, serviceGatewayName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._getAddressLocationsNext(resourceGroupName, serviceGatewayName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *getAddressLocationsPagingAll(resourceGroupName, serviceGatewayName, options) {
        for await (const page of this.getAddressLocationsPagingPage(resourceGroupName, serviceGatewayName, options)) {
            yield* page;
        }
    }
    /**
     * Get Services in service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    listServices(resourceGroupName, serviceGatewayName, options) {
        const iter = this.getServicesPagingAll(resourceGroupName, serviceGatewayName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.getServicesPagingPage(resourceGroupName, serviceGatewayName, options, settings);
            },
        };
    }
    async *getServicesPagingPage(resourceGroupName, serviceGatewayName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._getServices(resourceGroupName, serviceGatewayName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._getServicesNext(resourceGroupName, serviceGatewayName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *getServicesPagingAll(resourceGroupName, serviceGatewayName, options) {
        for await (const page of this.getServicesPagingPage(resourceGroupName, serviceGatewayName, options)) {
            yield* page;
        }
    }
    /**
     * Deletes the specified service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, serviceGatewayName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: { resourceGroupName, serviceGatewayName, options },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes the specified service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, serviceGatewayName, options) {
        const poller = await this.beginDelete(resourceGroupName, serviceGatewayName, options);
        return poller.pollUntilDone();
    }
    /**
     * Gets the specified service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    get(resourceGroupName, serviceGatewayName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, options }, getOperationSpec);
    }
    /**
     * Creates or updates a service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or update service gateway operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, serviceGatewayName, parameters, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: { resourceGroupName, serviceGatewayName, parameters, options },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Creates or updates a service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or update service gateway operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, serviceGatewayName, parameters, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, serviceGatewayName, parameters, options);
        return poller.pollUntilDone();
    }
    /**
     * Updates a service gateway tags.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to update service gateway tags.
     * @param options The options parameters.
     */
    updateTags(resourceGroupName, serviceGatewayName, parameters, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, parameters, options }, updateTagsOperationSpec);
    }
    /**
     * Gets all the service gateways in a subscription.
     * @param options The options parameters.
     */
    _listAll(options) {
        return this.client.sendOperationRequest({ options }, listAllOperationSpec);
    }
    /**
     * Gets all the service gateways in a resource group.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param options The options parameters.
     */
    _list(resourceGroupName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, options }, listOperationSpec);
    }
    /**
     * Creates or updates address locations within the service gateway.
     *
     * The request supports both full and partial update modes at two levels: location and address.
     *
     * Full update replaces all existing data.
     *
     * Partial update modifies only the specified entries:
     *
     * For location-level partial updates, if no address is provided, the existing address will be deleted.
     *
     * For address-level partial updates, if no services are provided, the existing services will be
     * considered for deletion.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or updates address locations in service gateway
     *                   operation.
     * @param options The options parameters.
     */
    async beginUpdateAddressLocations(resourceGroupName, serviceGatewayName, parameters, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: { resourceGroupName, serviceGatewayName, parameters, options },
            spec: updateAddressLocationsOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Creates or updates address locations within the service gateway.
     *
     * The request supports both full and partial update modes at two levels: location and address.
     *
     * Full update replaces all existing data.
     *
     * Partial update modifies only the specified entries:
     *
     * For location-level partial updates, if no address is provided, the existing address will be deleted.
     *
     * For address-level partial updates, if no services are provided, the existing services will be
     * considered for deletion.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or updates address locations in service gateway
     *                   operation.
     * @param options The options parameters.
     */
    async beginUpdateAddressLocationsAndWait(resourceGroupName, serviceGatewayName, parameters, options) {
        const poller = await this.beginUpdateAddressLocations(resourceGroupName, serviceGatewayName, parameters, options);
        return poller.pollUntilDone();
    }
    /**
     * Creates, updates, or deletes services within the service gateway.
     * The request supports both full and partial update modes at the service level.
     *
     * Full update replaces all existing services with the new list provided in the request.
     * Partial update modifies only the specified services.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or updates services in service gateway
     *                   operation.
     * @param options The options parameters.
     */
    async beginUpdateServices(resourceGroupName, serviceGatewayName, parameters, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: { resourceGroupName, serviceGatewayName, parameters, options },
            spec: updateServicesOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Creates, updates, or deletes services within the service gateway.
     * The request supports both full and partial update modes at the service level.
     *
     * Full update replaces all existing services with the new list provided in the request.
     * Partial update modifies only the specified services.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param parameters Parameters supplied to the create or updates services in service gateway
     *                   operation.
     * @param options The options parameters.
     */
    async beginUpdateServicesAndWait(resourceGroupName, serviceGatewayName, parameters, options) {
        const poller = await this.beginUpdateServices(resourceGroupName, serviceGatewayName, parameters, options);
        return poller.pollUntilDone();
    }
    /**
     * Get address locations in service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    _getAddressLocations(resourceGroupName, serviceGatewayName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, options }, getAddressLocationsOperationSpec);
    }
    /**
     * Get Services in service gateway.
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param options The options parameters.
     */
    _getServices(resourceGroupName, serviceGatewayName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, options }, getServicesOperationSpec);
    }
    /**
     * ListAllNext
     * @param nextLink The nextLink from the previous successful call to the ListAll method.
     * @param options The options parameters.
     */
    _listAllNext(nextLink, options) {
        return this.client.sendOperationRequest({ nextLink, options }, listAllNextOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, nextLink, options }, listNextOperationSpec);
    }
    /**
     * GetAddressLocationsNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param nextLink The nextLink from the previous successful call to the GetAddressLocations method.
     * @param options The options parameters.
     */
    _getAddressLocationsNext(resourceGroupName, serviceGatewayName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, nextLink, options }, getAddressLocationsNextOperationSpec);
    }
    /**
     * GetServicesNext
     * @param resourceGroupName The name of the resource group. The name is case insensitive.
     * @param serviceGatewayName The name of the service gateway.
     * @param nextLink The nextLink from the previous successful call to the GetServices method.
     * @param options The options parameters.
     */
    _getServicesNext(resourceGroupName, serviceGatewayName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, serviceGatewayName, nextLink, options }, getServicesNextOperationSpec);
    }
}
exports.ServiceGatewaysImpl = ServiceGatewaysImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.ServiceGatewaysDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.ServiceGatewaysDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.ServiceGatewaysDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.ServiceGatewaysDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGateway,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGateway,
        },
        201: {
            bodyMapper: Mappers.ServiceGateway,
        },
        202: {
            bodyMapper: Mappers.ServiceGateway,
        },
        204: {
            bodyMapper: Mappers.ServiceGateway,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters82,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const updateTagsOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}",
    httpMethod: "PATCH",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGateway,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters1,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const listAllOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Network/serviceGateways",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGatewayListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.$host, Parameters.subscriptionId],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGatewayListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [Parameters.$host, Parameters.subscriptionId, Parameters.resourceGroupName1],
    headerParameters: [Parameters.accept],
    serializer,
};
const updateAddressLocationsOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}/updateAddressLocations",
    httpMethod: "POST",
    responses: {
        200: {
            headersMapper: Mappers.ServiceGatewaysUpdateAddressLocationsHeaders,
        },
        201: {
            headersMapper: Mappers.ServiceGatewaysUpdateAddressLocationsHeaders,
        },
        202: {
            headersMapper: Mappers.ServiceGatewaysUpdateAddressLocationsHeaders,
        },
        204: {
            headersMapper: Mappers.ServiceGatewaysUpdateAddressLocationsHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters83,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const updateServicesOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}/updateServices",
    httpMethod: "POST",
    responses: {
        200: {
            headersMapper: Mappers.ServiceGatewaysUpdateServicesHeaders,
        },
        201: {
            headersMapper: Mappers.ServiceGatewaysUpdateServicesHeaders,
        },
        202: {
            headersMapper: Mappers.ServiceGatewaysUpdateServicesHeaders,
        },
        204: {
            headersMapper: Mappers.ServiceGatewaysUpdateServicesHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters84,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const getAddressLocationsOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}/addressLocations",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GetServiceGatewayAddressLocationsResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getServicesOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/serviceGateways/{serviceGatewayName}/services",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GetServiceGatewayServicesResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listAllNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGatewayListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [Parameters.$host, Parameters.subscriptionId, Parameters.nextLink],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ServiceGatewayListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getAddressLocationsNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GetServiceGatewayAddressLocationsResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getServicesNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GetServiceGatewayServicesResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName1,
        Parameters.serviceGatewayName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
