// @azure/arm-network/dist/react-native/operations/configurationPolicyGroups.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
import { createHttpPoller } from "@azure/core-lro";
import { createLroSpec } from "../lroImpl.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing ConfigurationPolicyGroups operations. */
export class ConfigurationPolicyGroupsImpl {
    client;
    /**
     * Initialize a new instance of the class ConfigurationPolicyGroups class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists all the configurationPolicyGroups in a resource group for a vpnServerConfiguration.
     * @param resourceGroupName The resource group name of the VpnServerConfiguration.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param options The options parameters.
     */
    listByVpnServerConfiguration(resourceGroupName, vpnServerConfigurationName, options) {
        const iter = this.listByVpnServerConfigurationPagingAll(resourceGroupName, vpnServerConfigurationName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByVpnServerConfigurationPagingPage(resourceGroupName, vpnServerConfigurationName, options, settings);
            },
        };
    }
    async *listByVpnServerConfigurationPagingPage(resourceGroupName, vpnServerConfigurationName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listByVpnServerConfiguration(resourceGroupName, vpnServerConfigurationName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listByVpnServerConfigurationNext(resourceGroupName, vpnServerConfigurationName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listByVpnServerConfigurationPagingAll(resourceGroupName, vpnServerConfigurationName, options) {
        for await (const page of this.listByVpnServerConfigurationPagingPage(resourceGroupName, vpnServerConfigurationName, options)) {
            yield* page;
        }
    }
    /**
     * Creates a ConfigurationPolicyGroup if it doesn't exist else updates the existing one.
     * @param resourceGroupName The resource group name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param configurationPolicyGroupName The name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationPolicyGroupParameters Parameters supplied to create or update a
     *                                                    VpnServerConfiguration PolicyGroup.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, vpnServerConfigurationPolicyGroupParameters, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: {
                resourceGroupName,
                vpnServerConfigurationName,
                configurationPolicyGroupName,
                vpnServerConfigurationPolicyGroupParameters,
                options,
            },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Creates a ConfigurationPolicyGroup if it doesn't exist else updates the existing one.
     * @param resourceGroupName The resource group name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param configurationPolicyGroupName The name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationPolicyGroupParameters Parameters supplied to create or update a
     *                                                    VpnServerConfiguration PolicyGroup.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, vpnServerConfigurationPolicyGroupParameters, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, vpnServerConfigurationPolicyGroupParameters, options);
        return poller.pollUntilDone();
    }
    /**
     * Deletes a ConfigurationPolicyGroup.
     * @param resourceGroupName The resource group name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param configurationPolicyGroupName The name of the ConfigurationPolicyGroup.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: {
                resourceGroupName,
                vpnServerConfigurationName,
                configurationPolicyGroupName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes a ConfigurationPolicyGroup.
     * @param resourceGroupName The resource group name of the ConfigurationPolicyGroup.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param configurationPolicyGroupName The name of the ConfigurationPolicyGroup.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, options) {
        const poller = await this.beginDelete(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, options);
        return poller.pollUntilDone();
    }
    /**
     * Retrieves the details of a ConfigurationPolicyGroup.
     * @param resourceGroupName The resource group name of the VpnServerConfiguration.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param configurationPolicyGroupName The name of the ConfigurationPolicyGroup being retrieved.
     * @param options The options parameters.
     */
    get(resourceGroupName, vpnServerConfigurationName, configurationPolicyGroupName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            vpnServerConfigurationName,
            configurationPolicyGroupName,
            options,
        }, getOperationSpec);
    }
    /**
     * Lists all the configurationPolicyGroups in a resource group for a vpnServerConfiguration.
     * @param resourceGroupName The resource group name of the VpnServerConfiguration.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param options The options parameters.
     */
    _listByVpnServerConfiguration(resourceGroupName, vpnServerConfigurationName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vpnServerConfigurationName, options }, listByVpnServerConfigurationOperationSpec);
    }
    /**
     * ListByVpnServerConfigurationNext
     * @param resourceGroupName The resource group name of the VpnServerConfiguration.
     * @param vpnServerConfigurationName The name of the VpnServerConfiguration.
     * @param nextLink The nextLink from the previous successful call to the ListByVpnServerConfiguration
     *                 method.
     * @param options The options parameters.
     */
    _listByVpnServerConfigurationNext(resourceGroupName, vpnServerConfigurationName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, vpnServerConfigurationName, nextLink, options }, listByVpnServerConfigurationNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/vpnServerConfigurations/{vpnServerConfigurationName}/configurationPolicyGroups/{configurationPolicyGroupName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.VpnServerConfigurationPolicyGroup,
        },
        201: {
            bodyMapper: Mappers.VpnServerConfigurationPolicyGroup,
        },
        202: {
            bodyMapper: Mappers.VpnServerConfigurationPolicyGroup,
        },
        204: {
            bodyMapper: Mappers.VpnServerConfigurationPolicyGroup,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.vpnServerConfigurationPolicyGroupParameters,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.vpnServerConfigurationName,
        Parameters.configurationPolicyGroupName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/vpnServerConfigurations/{vpnServerConfigurationName}/configurationPolicyGroups/{configurationPolicyGroupName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        201: {},
        202: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.vpnServerConfigurationName,
        Parameters.configurationPolicyGroupName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/vpnServerConfigurations/{vpnServerConfigurationName}/configurationPolicyGroups/{configurationPolicyGroupName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.VpnServerConfigurationPolicyGroup,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.vpnServerConfigurationName,
        Parameters.configurationPolicyGroupName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByVpnServerConfigurationOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/vpnServerConfigurations/{vpnServerConfigurationName}/configurationPolicyGroups",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ListVpnServerConfigurationPolicyGroupsResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.vpnServerConfigurationName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByVpnServerConfigurationNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ListVpnServerConfigurationPolicyGroupsResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.vpnServerConfigurationName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
