// @azure/arm-network/dist/react-native/operations/ipamPools.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
import { createHttpPoller } from "@azure/core-lro";
import { createLroSpec } from "../lroImpl.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing IpamPools operations. */
export class IpamPoolsImpl {
    client;
    /**
     * Initialize a new instance of the class IpamPools class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets list of Pool resources at Network Manager level.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkManagerName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkManagerName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkManagerName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkManagerName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkManagerName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkManagerName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkManagerName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkManagerName, options)) {
            yield* page;
        }
    }
    /**
     * List Associated Resource in the Pool.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    listAssociatedResources(resourceGroupName, networkManagerName, poolName, options) {
        const iter = this.listAssociatedResourcesPagingAll(resourceGroupName, networkManagerName, poolName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listAssociatedResourcesPagingPage(resourceGroupName, networkManagerName, poolName, options, settings);
            },
        };
    }
    async *listAssociatedResourcesPagingPage(resourceGroupName, networkManagerName, poolName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._listAssociatedResources(resourceGroupName, networkManagerName, poolName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listAssociatedResourcesNext(resourceGroupName, networkManagerName, poolName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listAssociatedResourcesPagingAll(resourceGroupName, networkManagerName, poolName, options) {
        for await (const page of this.listAssociatedResourcesPagingPage(resourceGroupName, networkManagerName, poolName, options)) {
            yield* page;
        }
    }
    /**
     * Gets list of Pool resources at Network Manager level.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkManagerName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, options }, listOperationSpec);
    }
    /**
     * Creates/Updates the Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName IP Address Manager Pool resource name.
     * @param body Pool resource object to create/update.
     * @param options The options parameters.
     */
    async beginCreate(resourceGroupName, networkManagerName, poolName, body, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: { resourceGroupName, networkManagerName, poolName, body, options },
            spec: createOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "azure-async-operation",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Creates/Updates the Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName IP Address Manager Pool resource name.
     * @param body Pool resource object to create/update.
     * @param options The options parameters.
     */
    async beginCreateAndWait(resourceGroupName, networkManagerName, poolName, body, options) {
        const poller = await this.beginCreate(resourceGroupName, networkManagerName, poolName, body, options);
        return poller.pollUntilDone();
    }
    /**
     * Updates the specific Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName IP Address Manager Pool resource name.
     * @param options The options parameters.
     */
    update(resourceGroupName, networkManagerName, poolName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, poolName, options }, updateOperationSpec);
    }
    /**
     * Gets the specific Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkManagerName, poolName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, poolName, options }, getOperationSpec);
    }
    /**
     * Delete the Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, networkManagerName, poolName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: { resourceGroupName, networkManagerName, poolName, options },
            spec: deleteOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Delete the Pool resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, networkManagerName, poolName, options) {
        const poller = await this.beginDelete(resourceGroupName, networkManagerName, poolName, options);
        return poller.pollUntilDone();
    }
    /**
     * Get the Pool Usage.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    getPoolUsage(resourceGroupName, networkManagerName, poolName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, poolName, options }, getPoolUsageOperationSpec);
    }
    /**
     * List Associated Resource in the Pool.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param options The options parameters.
     */
    _listAssociatedResources(resourceGroupName, networkManagerName, poolName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, poolName, options }, listAssociatedResourcesOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkManagerName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, nextLink, options }, listNextOperationSpec);
    }
    /**
     * ListAssociatedResourcesNext
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param poolName Pool resource name.
     * @param nextLink The nextLink from the previous successful call to the ListAssociatedResources
     *                 method.
     * @param options The options parameters.
     */
    _listAssociatedResourcesNext(resourceGroupName, networkManagerName, poolName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, poolName, nextLink, options }, listAssociatedResourcesNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.IpamPoolList,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.skipToken,
        Parameters.skip,
        Parameters.top1,
        Parameters.sortKey,
        Parameters.sortValue,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.IpamPool,
        },
        201: {
            bodyMapper: Mappers.IpamPool,
        },
        202: {
            bodyMapper: Mappers.IpamPool,
        },
        204: {
            bodyMapper: Mappers.IpamPool,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    requestBody: Parameters.body,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType, Parameters.ifMatch],
    mediaType: "json",
    serializer,
};
const updateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}",
    httpMethod: "PATCH",
    responses: {
        200: {
            bodyMapper: Mappers.IpamPool,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    requestBody: Parameters.body1,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType, Parameters.ifMatch],
    mediaType: "json",
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.IpamPool,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.IpamPoolsDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.IpamPoolsDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.IpamPoolsDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.IpamPoolsDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept, Parameters.ifMatch],
    serializer,
};
const getPoolUsageOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}/getPoolUsage",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: Mappers.PoolUsage,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listAssociatedResourcesOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/ipamPools/{poolName}/listAssociatedResources",
    httpMethod: "POST",
    responses: {
        200: {
            bodyMapper: Mappers.PoolAssociationList,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.IpamPoolList,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkManagerName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listAssociatedResourcesNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.PoolAssociationList,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkManagerName1,
        Parameters.poolName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
