// @azure/arm-network/dist/commonjs/operations/networkSecurityPerimeterLoggingConfigurations.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkSecurityPerimeterLoggingConfigurationsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing NetworkSecurityPerimeterLoggingConfigurations operations. */
class NetworkSecurityPerimeterLoggingConfigurationsImpl {
    client;
    /**
     * Initialize a new instance of the class NetworkSecurityPerimeterLoggingConfigurations class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists the NSP logging configuration.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkSecurityPerimeterName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkSecurityPerimeterName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkSecurityPerimeterName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkSecurityPerimeterName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkSecurityPerimeterName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkSecurityPerimeterName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, options)) {
            yield* page;
        }
    }
    /**
     * Gets the NSP logging configuration.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param loggingConfigurationName The name of the NSP logging configuration. Accepts 'instance' as
     *                                 name.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkSecurityPerimeterName, loggingConfigurationName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            loggingConfigurationName,
            options,
        }, getOperationSpec);
    }
    /**
     * Creates or updates NSP logging configuration.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param loggingConfigurationName The name of the NSP logging configuration. Accepts 'instance' as
     *                                 name.
     * @param parameters Parameters that hold the NspLoggingConfiguration to be created/updated.
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, networkSecurityPerimeterName, loggingConfigurationName, parameters, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            loggingConfigurationName,
            parameters,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Deletes an NSP Logging configuration.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param loggingConfigurationName The name of the NSP logging configuration. Accepts 'instance' as
     *                                 name.
     * @param options The options parameters.
     */
    delete(resourceGroupName, networkSecurityPerimeterName, loggingConfigurationName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            loggingConfigurationName,
            options,
        }, deleteOperationSpec);
    }
    /**
     * Lists the NSP logging configuration.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkSecurityPerimeterName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, options }, listOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkSecurityPerimeterName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, nextLink, options }, listNextOperationSpec);
    }
}
exports.NetworkSecurityPerimeterLoggingConfigurationsImpl = NetworkSecurityPerimeterLoggingConfigurationsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/loggingConfigurations/{loggingConfigurationName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLoggingConfiguration,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.loggingConfigurationName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/loggingConfigurations/{loggingConfigurationName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.NspLoggingConfiguration,
        },
        201: {
            bodyMapper: Mappers.NspLoggingConfiguration,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters53,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.loggingConfigurationName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/loggingConfigurations/{loggingConfigurationName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.loggingConfigurationName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/loggingConfigurations",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLoggingConfigurationListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLoggingConfigurationListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkSecurityPerimeterName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
