// @azure/arm-network/dist/commonjs/operations/reachabilityAnalysisRuns.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReachabilityAnalysisRunsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing ReachabilityAnalysisRuns operations. */
class ReachabilityAnalysisRunsImpl {
    client;
    /**
     * Initialize a new instance of the class ReachabilityAnalysisRuns class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Gets list of Reachability Analysis Runs.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkManagerName, workspaceName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkManagerName, workspaceName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkManagerName, workspaceName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkManagerName, workspaceName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkManagerName, workspaceName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkManagerName, workspaceName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkManagerName, workspaceName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkManagerName, workspaceName, options)) {
            yield* page;
        }
    }
    /**
     * Gets list of Reachability Analysis Runs.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkManagerName, workspaceName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkManagerName, workspaceName, options }, listOperationSpec);
    }
    /**
     * Gets Reachability Analysis Run.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisRunName Reachability Analysis Run name.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisRunName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            reachabilityAnalysisRunName,
            options,
        }, getOperationSpec);
    }
    /**
     * Creates Reachability Analysis Runs.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisRunName Reachability Analysis Run name.
     * @param body Analysis Run resource object to create/update.
     * @param options The options parameters.
     */
    create(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisRunName, body, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            reachabilityAnalysisRunName,
            body,
            options,
        }, createOperationSpec);
    }
    /**
     * Deletes Reachability Analysis Run.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisRunName Reachability Analysis Run name.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisRunName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                networkManagerName,
                workspaceName,
                reachabilityAnalysisRunName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes Reachability Analysis Run.
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param reachabilityAnalysisRunName Reachability Analysis Run name.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisRunName, options) {
        const poller = await this.beginDelete(resourceGroupName, networkManagerName, workspaceName, reachabilityAnalysisRunName, options);
        return poller.pollUntilDone();
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkManagerName The name of the network manager.
     * @param workspaceName Workspace name.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkManagerName, workspaceName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkManagerName,
            workspaceName,
            nextLink,
            options,
        }, listNextOperationSpec);
    }
}
exports.ReachabilityAnalysisRunsImpl = ReachabilityAnalysisRunsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisRuns",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisRunListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.skipToken,
        Parameters.skip,
        Parameters.top1,
        Parameters.sortKey,
        Parameters.sortValue,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisRuns/{reachabilityAnalysisRunName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisRun,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisRunName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisRuns/{reachabilityAnalysisRunName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisRun,
        },
        201: {
            bodyMapper: Mappers.ReachabilityAnalysisRun,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    requestBody: Parameters.body4,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisRunName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkManagers/{networkManagerName}/verifierWorkspaces/{workspaceName}/reachabilityAnalysisRuns/{reachabilityAnalysisRunName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.ReachabilityAnalysisRunsDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.ReachabilityAnalysisRunsDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.ReachabilityAnalysisRunsDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.ReachabilityAnalysisRunsDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
        Parameters.reachabilityAnalysisRunName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.ReachabilityAnalysisRunListResult,
        },
        default: {
            bodyMapper: Mappers.CommonErrorResponse,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkManagerName1,
        Parameters.workspaceName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
