// @azure/arm-network/dist/browser/operations/networkSecurityPerimeterLinks.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { setContinuationToken } from "../pagingHelper.js";
import * as coreClient from "@azure/core-client";
import * as Mappers from "../models/mappers.js";
import * as Parameters from "../models/parameters.js";
import { createHttpPoller } from "@azure/core-lro";
import { createLroSpec } from "../lroImpl.js";
/// <reference lib="esnext.asynciterable" />
/** Class containing NetworkSecurityPerimeterLinks operations. */
export class NetworkSecurityPerimeterLinksImpl {
    client;
    /**
     * Initialize a new instance of the class NetworkSecurityPerimeterLinks class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * Lists the NSP Link resources in the specified network security perimeter.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param options The options parameters.
     */
    list(resourceGroupName, networkSecurityPerimeterName, options) {
        const iter = this.listPagingAll(resourceGroupName, networkSecurityPerimeterName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings?.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, options, settings);
            },
        };
    }
    async *listPagingPage(resourceGroupName, networkSecurityPerimeterName, options, settings) {
        let result;
        let continuationToken = settings?.continuationToken;
        if (!continuationToken) {
            result = await this._list(resourceGroupName, networkSecurityPerimeterName, options);
            const page = result.value || [];
            continuationToken = result.nextLink;
            setContinuationToken(page, continuationToken);
            yield page;
        }
        while (continuationToken) {
            result = await this._listNext(resourceGroupName, networkSecurityPerimeterName, continuationToken, options);
            continuationToken = result.nextLink;
            const page = result.value || [];
            setContinuationToken(page, continuationToken);
            yield page;
        }
    }
    async *listPagingAll(resourceGroupName, networkSecurityPerimeterName, options) {
        for await (const page of this.listPagingPage(resourceGroupName, networkSecurityPerimeterName, options)) {
            yield* page;
        }
    }
    /**
     * Gets the specified NSP link resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param linkName The name of the NSP link.
     * @param options The options parameters.
     */
    get(resourceGroupName, networkSecurityPerimeterName, linkName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, linkName, options }, getOperationSpec);
    }
    /**
     * Creates or updates NSP link resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param linkName The name of the NSP link.
     * @param parameters Parameters that hold the NspLink resource to be created/updated.
     * @param options The options parameters.
     */
    createOrUpdate(resourceGroupName, networkSecurityPerimeterName, linkName, parameters, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            networkSecurityPerimeterName,
            linkName,
            parameters,
            options,
        }, createOrUpdateOperationSpec);
    }
    /**
     * Deletes an NSP Link resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param linkName The name of the NSP link.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, networkSecurityPerimeterName, linkName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            let currentRawResponse = undefined;
            const providedCallback = args.options?.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback?.(rawResponse, flatResponse);
            };
            const updatedArgs = {
                ...args,
                options: {
                    ...args.options,
                    onResponse: callback,
                },
            };
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = createLroSpec({
            sendOperationFn,
            args: {
                resourceGroupName,
                networkSecurityPerimeterName,
                linkName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await createHttpPoller(lro, {
            restoreFrom: options?.resumeFrom,
            intervalInMs: options?.updateIntervalInMs,
            resourceLocationConfig: "location",
        });
        await poller.poll();
        return poller;
    }
    /**
     * Deletes an NSP Link resource.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param linkName The name of the NSP link.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, networkSecurityPerimeterName, linkName, options) {
        const poller = await this.beginDelete(resourceGroupName, networkSecurityPerimeterName, linkName, options);
        return poller.pollUntilDone();
    }
    /**
     * Lists the NSP Link resources in the specified network security perimeter.
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param options The options parameters.
     */
    _list(resourceGroupName, networkSecurityPerimeterName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, options }, listOperationSpec);
    }
    /**
     * ListNext
     * @param resourceGroupName The name of the resource group.
     * @param networkSecurityPerimeterName The name of the network security perimeter.
     * @param nextLink The nextLink from the previous successful call to the List method.
     * @param options The options parameters.
     */
    _listNext(resourceGroupName, networkSecurityPerimeterName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, networkSecurityPerimeterName, nextLink, options }, listNextOperationSpec);
    }
}
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/links/{linkName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLink,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.linkName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/links/{linkName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.NspLink,
        },
        201: {
            bodyMapper: Mappers.NspLink,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.parameters52,
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.linkName1,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/links/{linkName}",
    httpMethod: "DELETE",
    responses: {
        200: {},
        201: {},
        202: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
        Parameters.linkName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkSecurityPerimeters/{networkSecurityPerimeterName}/links",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLinkListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion, Parameters.top, Parameters.skipToken1],
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.networkSecurityPerimeterName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.NspLinkListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.resourceGroupName,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.networkSecurityPerimeterName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
